source/auth/auth.rst

```diff
@@ -1175,70 +1175,148 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
 .. note::
     Five minutes was chosen based on the AWS documentation for `IAM roles for EC2 <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html>`_ : "We make new credentials available at least five minutes before the expiration of the old credentials". The intent is to have some buffer between when the driver fetches the credentials and when the server verifies them.
 
+
 MONGODB-OIDC
 ~~~~~~~~~~~~
 
 :since: 7.0 Enterprise
 
-MONGODB-OIDC authenticates using an `OIDC <https://openid.net/specs/openid-connect-core-1_0.html>`_ access tokens.  Drivers MUST support
-both Callback-driven OIDC and Automatic OIDC Authentication for AWS.
+MONGODB-OIDC authenticates using an `OpenID Connect (OIDC)
+<https://openid.net/specs/openid-connect-core-1_0.html>`_ access token. Most
+OIDC Identity Providers (*IdP*) support two identity federation flows:
+
+- Workload Identity Federation - Also called the "machine authentication flow",
+  this flow is intended to authorize access for software and services without
+  requiring human interaction.
+
+  TODO: Is there a 1:1 corresponding OIDC spec flow?
+
+- Workforce Identity Federation - Also called the "human authentication flow",
+  this flow is intended to authorize access for end users and typically requires
+  human interaction to complete (e.g. clicking a button in a browser).
+
+  TODO: Is there a 1:1 corresponding OIDC spec flow?
+
+Drivers MUST implement the Workload Identity Federation flow. Drivers MAY
+implement the Workforce Identity Federation flow.
+
+
+Workload Identity Federation
+----------------------------
+
+`MongoCredential`_ Properties
+`````````````````````````````
+username
+    MUST NOT be specified.
+
+source
+    MUST be "$external". Defaults to ``$external``.
+
+password
+    MUST NOT be specified.
+
+mechanism
+    MUST be "MONGODB-OIDC"
+
+mechanism_properties
+    PROVIDER_NAME
+        Drivers MUST allow the user to specify the name of the OIDC service to
+        use to obtain credentials. MUST be one of ["aws"].
+
+Interfaces
+``````````
+
+Custom Callback
+```````````````
+Drivers MUST provide a way for the callback to be either
+automatically canceled, or to cancel itself.  This can be as a timeout argument
+to the callback, a cancellation context passed to the callback, or some other
+language-appropriate mechanism.  The timeout deadline MUST be the same as the
+connection establishment deadline (TODO: Does that make sense?).
+
+Callbacks can be synchronous and/or asynchronous, depending on the driver
+and/or language.  Asynchronous callbacks should be preferred when other
+operations in the driver use asynchronous functions.
+
+The driver MUST pass the following information to the callback:
+
+- ``timeoutMS``: A timeout, in milliseconds, deadline, or ``timeoutContext``.
+- ``mechanismProperties``: All key-value pairs from the
```

Comment:
This seems odd.  The driver will pass `["PROVIDER_NAME": "aws"]` but will have no way of enforcing that "aws" is actually used by the callback. Also, isn't the callback itself specified as an auth mechanism property?

Seems like if a callback is provided, it shouldn't need any of the mechanism properties and therefore not should be required when a callback is provided.

Reply:
My initial intent was to define an API that would satisfy either a custom machine-to-machine callback or a built-in callback (e.g. the future built-in Azure OIDC logic, which will need extra configs to be provided by `authMechanismProperties`). On reflection, that seems overly prescriptive of the implementation rather than just the configuration API. I also intended `authMechanismProperties` to communicate additional configuration to the custom callback, but I realized that drivers validate the keys in `authMechanismProperties`, so passing arbitrary config values isn't possible.

As far as specifying the callback name in `authMechanismProperties`, I'm currently planning to drop that from the spec. I haven't been able to discover a use case for it and implementing that logic would be unconventional or nonsensical in many driver languages.

Reply:
Updated to not pass the mechanism properties to the callback and require that drivers enforce that `PROVIDER_NAME` and a custom callback cannot be specified at the same time. Also updated to support either specifying the callback via a `MongoClientConfig` property or a mechanism property, depending on which is supported or idiomatic for each driver.